core/payment/src/processor.rs

```diff
@@ -306,8 +311,117 @@ impl DriverRegistry {
 const DB_LOCK_TIMEOUT: Duration = Duration::from_secs(30);
 const REGISTRY_LOCK_TIMEOUT: Duration = Duration::from_secs(30);
 
+struct TimedMutex {
+    mutex: Mutex<DbExecutor>,
+    sender: Option<UnboundedSender<TimedMutexTaskMessage>>,
+    counter_task: Option<JoinHandle<()>>,
+}
+
+use tokio::sync::MutexGuard;
+
+enum TimedMutexTaskMessage {
+    Start(String),
+    Finish,
+}
+
+struct TimedMutexGuard<'a> {
+    mutex_guard: MutexGuard<'a, DbExecutor>,
+    sender: &'a Option<UnboundedSender<TimedMutexTaskMessage>>,
+}
+
+impl Drop for TimedMutexGuard<'_> {
+    fn drop(&mut self) {
+        if let Some(sender) = &self.sender {
+            if let Err(e) = sender.send(TimedMutexTaskMessage::Finish) {
+                log::error!("Cannot send fininsh to counter task {e}");
+            }
+        }
+    }
+}
+
+impl<'a> Deref for TimedMutexGuard<'a> {
+    type Target = MutexGuard<'a, DbExecutor>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.mutex_guard
+    }
+}
+
+impl TimedMutex {
+    fn new(db: DbExecutor) -> Self {
+        let (sender, mut receiver) =
+            tokio::sync::mpsc::unbounded_channel::<TimedMutexTaskMessage>();
+
+        let counter_task = tokio::spawn(async move {
```

Comment:
I think there is a slight problem - tokio swallows panic and we ignore it.
I think we should check counter_task.[is_finished()](https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#method.is_finished) when locking the mutex and log/abort/... when locking the mutex if it's in an unexpected state.